### PR TITLE
Vision Fixes

### PR DESCRIPTION
**Pull request by Strflightmight09**

### CODE REVIEW

File: src/main/java/frc/trigon/robot/constants/CameraConstants.java [hunk 1]
```diff
@@ -3,6 +3,9 @@
 import frc.trigon.robot.misc.objectdetectioncamera.ObjectDetectionCamera;
 
 public class CameraConstants {
+    public static final double
+            TRANSLATIONS_STD_EXPONENT = 0.005,
+            THETA_STD_EXPONENT = 0.01;
```
Comment (levyishai) [inline src/main/java/frc/trigon/robot/constants/CameraConstants.java:8]: these should be passed in the constructor

File: src/main/java/frc/trigon/robot/poseestimation/apriltagcamera/AprilTagCamera.java [hunk 2]
```diff
@@ -202,4 +194,18 @@ private void logUsedTags() {
...
+
+    private void logEstimatedRobotPose() {
+        if (!inputs.hasResult || inputs.distanceFromBestTag == 0 || robotPose == null)
+            Logger.recordOutput("Poses/Robot/" + name + "Pose", AprilTagCameraConstants.EMPTY_POSE_LIST);
+        else
+            Logger.recordOutput("Poses/Robot/" + name + "Pose", robotPose);
+    }
+
+    private void logSolvePNPPose() {
+        if (!inputs.hasResult || inputs.distanceFromBestTag == 0 || robotPose == null)
```
Comment (levyishai) [inline src/main/java/frc/trigon/robot/poseestimation/apriltagcamera/AprilTagCamera.java]: why check this twice?

File: src/main/java/frc/trigon/robot/poseestimation/apriltagcamera/AprilTagCamera.java [hunk 3]
```diff
@@ -67,21 +64,15 @@ public double getLatestResultTimestampSeconds() {
         return inputs.latestResultTimestampSeconds;
     }
 
-    public void setGyroHeadingToSolvePNPHeading() {
-        final Translation2d currentTranslation = RobotContainer.POSE_ESTIMATOR.getCurrentPose().getTranslation();
-        final Rotation2d solvePNPHeading = inputs.cameraSolvePNPPose.getRotation().toRotation2d();
-        RobotContainer.POSE_ESTIMATOR.resetPose(new Pose2d(currentTranslation, solvePNPHeading));
-    }
```
Comment (levyishai) [inline src/main/java/frc/trigon/robot/poseestimation/apriltagcamera/AprilTagCamera.java:74]: implement this somewhere, probably pose estimator (obviously not as simple as copy paste, you should reset to the camera with the smallest distance from best tag)

**Review by levyishai — CHANGES_REQUESTED** [review]

**Review by levyishai — CHANGES_REQUESTED** [review]
Instead of tossing an error down the drain every time I see it in hopes it'll fix itself, I've found the problem that was crashing the code: `AprilTagPhotonCameraIO.java:99`

File: src/main/java/frc/trigon/robot/poseestimation/apriltagcamera/AprilTagCamera.java [hunk 4]
```diff
@@ -202,4 +207,18 @@ private void logUsedTags() {
             usedTagPoses[i] = FieldConstants.TAG_ID_TO_POSE.get(inputs.visibleTagIDs[i]);
         Logger.recordOutput("UsedTags/" + this.getName(), usedTagPoses);
     }
+
+    private void logEstimatedRobotPose() {
+        if (!inputs.hasResult || inputs.distanceFromBestTag == 0 || robotPose == null)
```
Comment (levyishai) [inline src/main/java/frc/trigon/robot/poseestimation/apriltagcamera/AprilTagCamera.java]: This `if` should really happen once

File: src/main/java/frc/trigon/robot/poseestimation/poseestimator/PoseEstimator.java [hunk 5]
```diff
@@ -80,6 +80,17 @@ public void updatePoseEstimatorStates(SwerveDriveWheelPositions[] swerveWheelPos
...
     }
 
+    public void setGyroHeadingToBestSolvePNPHeading() {
+        int closestCameraToTag = 0;
+        for (int i = 0; i < aprilTagCameras.length; i++) {
+            if (aprilTagCameras[i].getDistanceToBestTagMeters() < aprilTagCameras[closestCameraToTag].getDistanceToBestTagMeters())
+                closestCameraToTag = i;
+        }
+
+        final Rotation2d bestRobotHeading = aprilTagCameras[closestCameraToTag].getEstimatedRobotPose().getRotation();
```
Comment (levyishai) [inline src/main/java/frc/trigon/robot/poseestimation/poseestimator/PoseEstimator.java]: This isn't necessarily solvepnp

**Review by levyishai — CHANGES_REQUESTED** [review]

File: src/main/java/frc/trigon/robot/poseestimation/apriltagcamera/AprilTagCamera.java [hunk 6]
```diff
@@ -202,4 +216,15 @@ private void logUsedTags() {
             usedTagPoses[i] = FieldConstants.TAG_ID_TO_POSE.get(inputs.visibleTagIDs[i]);
         Logger.recordOutput("UsedTags/" + this.getName(), usedTagPoses);
     }
+
+    private void logEstimatedRobotPose() {
+        if (robotPose == null)
+            Logger.recordOutput("Poses/Robot/" + name + "Pose", AprilTagCameraConstants.EMPTY_POSE_LIST);
+        else
+            Logger.recordOutput("Poses/Robot/" + name + "Pose", robotPose);
```
Comment (levyishai) [inline src/main/java/frc/trigon/robot/poseestimation/apriltagcamera/AprilTagCamera.java]: the robotPose == null can be in the other if
Add a slash after name

File: src/main/java/frc/trigon/robot/poseestimation/apriltagcamera/AprilTagCamera.java [hunk 7]
```diff
@@ -202,4 +216,15 @@ private void logUsedTags() {
...
+    }
+
+    private void logSolvePNPPose() {
+        Logger.recordOutput("Poses/Robot/" + name + "SolvePNPPose", inputs.cameraSolvePNPPose.plus(robotCenterToCamera.inverse()));
+    }
```
Comment (levyishai) [inline src/main/java/frc/trigon/robot/poseestimation/apriltagcamera/AprilTagCamera.java]: Add  a slash after name

File: src/main/java/frc/trigon/robot/poseestimation/apriltagcamera/AprilTagCamera.java [hunk 8]
```diff
@@ -184,11 +192,17 @@ private boolean isWithinBestTagRangeForSolvePNP() {
...
+        if (!FieldConstants.TAG_ID_TO_POSE.isEmpty())
+            logUsedTags();
+        if (!inputs.hasResult || inputs.distanceFromBestTag == 0) {
+            logEstimatedRobotPose();
+            logSolvePNPPose();
+        } else {
             Logger.recordOutput("Poses/Robot/" + name + "Pose", AprilTagCameraConstants.EMPTY_POSE_LIST);
-        else
-            Logger.recordOutput("Poses/Robot/" + name + "Pose", robotPose);
+            Logger.recordOutput("Poses/Robot/" + name + "SolvePNPPose", AprilTagCameraConstants.EMPTY_POSE_LIST);
```
Comment (levyishai) [inline src/main/java/frc/trigon/robot/poseestimation/apriltagcamera/AprilTagCamera.java]: Add a slash after name

**Review by levyishai — CHANGES_REQUESTED** [review]

File: src/main/java/frc/trigon/robot/poseestimation/apriltagcamera/AprilTagCamera.java [hunk 9]
```diff
@@ -59,6 +65,10 @@ public Pose2d getEstimatedRobotPose() {
         return robotPose;
     }
 
+    public Rotation2d getSolvePNPHeading() {
+        return inputs.cameraSolvePNPPose.toPose2d().getRotation();
+    }
```
Comment (levyishai) [inline src/main/java/frc/trigon/robot/poseestimation/apriltagcamera/AprilTagCamera.java]: will this return correctly?

File: src/main/java/frc/trigon/robot/poseestimation/poseestimator/PoseEstimator.java [hunk 10]
```diff
@@ -80,6 +80,17 @@ public void updatePoseEstimatorStates(SwerveDriveWheelPositions[] swerveWheelPos
...
+        int closestCameraToTag = 0;
+        for (int i = 0; i < aprilTagCameras.length; i++) {
+            if (aprilTagCameras[i].getDistanceToBestTagMeters() < aprilTagCameras[closestCameraToTag].getDistanceToBestTagMeters())
+                closestCameraToTag = i;
+        }
+
+        final Rotation2d bestRobotHeading = aprilTagCameras[closestCameraToTag].getSolvePNPHeading();
+        resetPose(new Pose2d(getCurrentPose().getTranslation(), bestRobotHeading));
+    }
+
```
Comment (levyishai) [inline src/main/java/frc/trigon/robot/poseestimation/poseestimator/PoseEstimator.java:93]: implement this button

**Review by levyishai — CHANGES_REQUESTED** [review]

File: src/main/java/frc/trigon/robot/commands/CommandConstants.java [hunk 11]
```diff
@@ -45,6 +45,7 @@ public class CommandConstants {
                     () -> calculateRotationStickAxisValue(DRIVER_CONTROLLER.getRightX())
             ),
             RESET_HEADING_COMMAND = new InstantCommand(() -> RobotContainer.POSE_ESTIMATOR.resetPose(changeRotation(new MirrorablePose2d(RobotContainer.POSE_ESTIMATOR.getCurrentPose(), false), new Rotation2d()).get())),
+            SET_GYRO_HEADING_TO_SOLVE_PNP_HEADING_COMMAND = new InstantCommand(RobotContainer.POSE_ESTIMATOR::setGyroHeadingToBestSolvePNPHeading),
             SELF_RELATIVE_DRIVE_FROM_DPAD_COMMAND = SwerveCommands.getClosedLoopSelfRelativeDriveCommand(
```
Comment (levyishai) [inline src/main/java/frc/trigon/robot/commands/CommandConstants.java:49]: Run when disabled

**Review by levyishai — CHANGES_REQUESTED** [review]

**Review by levyishai — APPROVED** [review]

**Review by Nummun14 — APPROVED** [review]

**Pull request merged**